File: 2016/python/day02.py
file = open("02.txt", "r")

keypad = [["1", "2", "3"],
          ["4", "5", "6"],
          ["7", "8", "9"]]

x = y = 1
code = ""
instructions = []
for line in file:
    instructions.append(line.strip())
file.close()

for i in instructions:
    for char in i:
        if char == "U": y = y - 1 if y > 0 else 0
        elif char == "D": y = y + 1 if y < 2 else 2
        elif char == "L": x = x - 1 if x > 0 else 0
        elif char == "R": x = x + 1 if x < 2 else 2
    code += keypad[y][x]

print("Code (puzzle part 1):", code)

keypad = [[" ", " ", "1", " ", " "],
          [" ", "2", "3", "4", " "],
          ["5", "6", "7", "8", "9"],
          [" ", "A", "B", "C", " "],
          [" ", " ", "D", " ", " "]]

x, y = 0, 2
code = ""

for i in instructions:
    for char in i:
        if char == "U": y = y if (y + x) == 2 or (y - x) == -2 else y - 1
        elif char == "D": y = y if (y + x) == 6 or (y - x) == 2 else y + 1
        elif char == "L": x = x if (y + x) == 2 or (y - x) == 2 else x - 1
        elif char == "R": x = x if (y + x) == 6 or (y - x) == -2 else x + 1
    code += keypad[y][x]

print("Code (puzzle part 2):", code)

File: 2019/lua/day02.lua
local code, mem = {}, {}

local function get(ptr) return mem[ptr] or code[ptr] end

local function run()
    local ptr = 0
    while get(ptr) ~= 99 do
        if get(ptr) == 1 then mem[get(ptr + 3)] = get(get(ptr + 1)) + get(get(ptr + 2))
        elseif get(ptr) == 2 then mem[get(ptr + 3)] = get(get(ptr + 1)) * get(get(ptr + 2)) end
        ptr = ptr + 4
    end
end

local function nvsearch(target)
    for noun = 0, 99 do
        for verb = 0, 99 do
            mem = {noun, verb}; run()
            
            if get(0) == target then return 100 * noun + verb end
        end end end

local n = 0
for item in string.gmatch(io.open("../__in/02.txt", "r"):read(), "[%d%-]+") do code[n] = tonumber(item); n = n + 1 end

mem = {12, 2}; run()
print(string.format("Part 1: %d", get(0)))
print(string.format("Part 2: %d", nvsearch(19690720)))